storages/backends/gcloud.py

```diff
@@ -0,0 +1,226 @@
+    bucket_acl = setting('GS_BUCKET_ACL', default_acl)
+    file_name_charset = setting('GS_FILE_NAME_CHARSET', 'utf-8')
+    file_overwrite = setting('GS_FILE_OVERWRITE', True)
+    # The max amount of memory a returned file can take up before being
+    # rolled over into a temporary file on disk. Default is 0: Do not roll over.
+    max_memory_size = setting('GS_MAX_MEMORY_SIZE', 0)
+
+    def __init__(self, **settings):
+        # check if some of the settings we've provided as class attributes
+        # need to be overwritten with values passed in here
+        for name, value in settings.items():
+            if hasattr(self, name):
+                setattr(self, name, value)
+
+        self._bucket = None
+        self._client = None
+
+    @property
+    def client(self):
+        if self._client is None:
+            self._client = self.client_class(
+                project=self.project_id,
+                credentials=self.credentials
+            )
+        return self._client
+
+    @property
+    def bucket(self):
+        if self._bucket is None:
+            self._bucket = self._get_or_create_bucket(self.bucket_name)
+        return self._bucket
+
+    def _get_or_create_bucket(self, name):
+        """
+        Retrieves a bucket if it exists, otherwise creates it.
+        """
+        try:
+            return self.client.get_bucket(name)
+        except self.not_found_exception:
+            if self.auto_create_bucket:
+                bucket = self.client.create_bucket(name)
+                bucket.acl.all().grant(self.bucket_acl)
+                bucket.acl.save()
+                return bucket
+            raise ImproperlyConfigured("Bucket %s does not exist. Buckets "
+                                       "can be automatically created by "
+                                       "setting GS_AUTO_CREATE_BUCKET to "
+                                       "``True``." % name)
+
+    def _clean_name(self, name):
+        """
+        Cleans the name so that Windows style paths work
+        """
+        return clean_name(name)
+
+    def _normalize_name(self, name):
+        """
+        Normalizes the name so that paths like /path/to/ignored/../something.txt
+        and ./file.txt work.  Note that clean_name adds ./ to some paths so
+        they need to be fixed here.
+        """
+        return safe_join('', name)
+
+    def _encode_name(self, name):
+        return smart_str(name, encoding=self.file_name_charset)
+
+    def _open(self, name, mode='rb'):
+        name = self._normalize_name(self._clean_name(name))
+        file_object = self.file_class(name, mode, self)
+        if not file_object.blob:
+            raise IOError('File does not exist: %s' % name)
+        return file_object
+
+    def _save(self, name, content):
+        cleaned_name = self._clean_name(name)
+        name = self._normalize_name(cleaned_name)
+        size = getattr(content, 'size')
```

Comment:
Why do this with `getattr`? If it doesn't exist this is going to fail with `AttributeError` anyway so it can just be inlined right i.e. `size=content.size`. Does the `.upload_from_file` not automatically infer size as well?